frameworks/moveos-stdlib/sources/event_queue.move

```diff
@@ -0,0 +1,217 @@
+        let tail_sequence_number = object::borrow(event_queue_obj).tail_sequence_number;
+        let now = timestamp::now_milliseconds();
+        let remove_sequence_number = tail_sequence_number;
+        while (remove_sequence_number < head_sequence_number){
+            if (object::contains_field(event_queue_obj, remove_sequence_number)){
+                let on_chain_event: &OnChainEvent<E> = object::borrow_field(event_queue_obj, remove_sequence_number);
+                if (now - on_chain_event.emit_time > EVENT_EXPIRE_TIME){
+                    let _event: OnChainEvent<E> = object::remove_field(event_queue_obj, remove_sequence_number);
+                }else{
+                    break
+                }
+            };
+            remove_sequence_number = remove_sequence_number + 1;
+            if (remove_sequence_number - tail_sequence_number >= REMOVE_EXPIRED_EVENT_BATCH_SIZE){
+                break
+            }
+        };
+        object::borrow_mut(event_queue_obj).tail_sequence_number = remove_sequence_number;
+    }
+
+    #[test_only]
+    struct TestEvent has copy, drop, store{
+        value: u64,
+    }
+
+    #[test]
+    fun test_event_queue_basic(){
+        let queue_name = std::string::utf8(b"test_event_queue");
+        let subscriber = subscribe<TestEvent>(queue_name);
+        emit(queue_name, TestEvent{value: 1});
+        let event = consume(&mut subscriber);
+        assert!(option::is_some(&event), 1000);
+        assert!(option::destroy_some(event).value == 1, 1001);
+        let event = consume(&mut subscriber);
+        assert!(option::is_none(&event), 1002);
+        emit(queue_name, TestEvent{value: 2});
+        let event = consume(&mut subscriber);
+        assert!(option::is_some(&event), 1003);
+        assert!(option::destroy_some(event).value == 2, 1004);
+        unsubscribe(subscriber);
+    }
+
+    #[test]
+    fun test_event_queue_two_subscriber(){
+        let queue_name = std::string::utf8(b"test_event_queue_two_subscriber");
+        let subscriber1 = subscribe<TestEvent>(queue_name);
+        emit(queue_name, TestEvent{value: 1});
+        let subscriber2 = subscribe<TestEvent>(queue_name);
+
+        let event1 = consume(&mut subscriber1);
+        let event2 = consume(&mut subscriber2);
+        
+        assert!(option::is_some(&event1), 1000);
+        assert!(option::destroy_some(event1).value == 1, 1001);
+        assert!(option::is_none(&event2), 1002);
+
+        emit(queue_name, TestEvent{value: 2});
+        let event1 = consume(&mut subscriber1);
+        let event2 = consume(&mut subscriber2);
+        
+        assert!(option::is_some(&event1), 1003);
+        assert!(option::destroy_some(event1).value == 2, 1004);
+        assert!(option::is_some(&event2), 1005);
+        assert!(option::destroy_some(event2).value == 2, 1006);
+
+        unsubscribe(subscriber1);
+        unsubscribe(subscriber2);
+    }
+
+    #[test]
+    fun test_event_queue_expired_events(){
```

Comment:
I think it would be better to use two consumers for testing here, one to trigger cleanup and the other to get expired events.

Reply:
The cleanup when emit event.

Reply:
Add too subscribers